### ISSUE TITLE
Disappearing links cause unbounded heap growth

### ISSUE THREAD

**Issue by civodul** [ivmai/bdwgc#182]
The attached file causes seemingly unbounded heap growth.  The problem disappears when the `GC_GENERAL_REGISTER_DISAPPEARING_LINK` call is commented out.

Is this "expected"?  How would suggest working around it?

Thanks,
Ludovic.

**Comment by civodul** [comment]
The reproducer:

```c
#define GC_DEBUG 1
#include <gc.h>
#include <stdlib.h>
#include <stdio.h>

int
main ()
{
  GC_INIT ();

  while (1)
    {
      unsigned int count = 777 * (random () % 1000);
      void **p = GC_MALLOC_ATOMIC (count * sizeof *p);

      for (unsigned int i = 0; i < count; i++)
	{
	  void *key = GC_MALLOC_ATOMIC (10);
	  p[i] = key;
	  GC_GENERAL_REGISTER_DISAPPEARING_LINK (&p[i], key); /* <- !!! */
	}

      static unsigned int loops = 0;
      if (loops++ % 10 == 0)
	printf ("iteration %4u, heap size = %li MiB\n",
		loops, GC_get_heap_size () / (1024 * 1024));
    }
}
```

**Comment by civodul** [comment]
A workaround (which is not always applicable, though) is to unregister all the disappearing links before `p` goes out of scope.

**Comment by ivmai** [comment]
I can't reproduce this on the latest master.
Host: cygwin/x86
Output:
iteration    1, heap size = 1 MiB
iteration   11, heap size = 13 MiB
iteration   21, heap size = 21 MiB
iteration   31, heap size = 21 MiB
iteration   41, heap size = 21 MiB
iteration   51, heap size = 21 MiB
iteration   61, heap size = 21 MiB
iteration   71, heap size = 21 MiB
iteration   81, heap size = 21 MiB
iteration   91, heap size = 21 MiB
iteration  101, heap size = 21 MiB

**Comment by ivmai** [comment]
@civodul Let me know the configuration (and host), if this is still reproducible?

**Comment by ivmai** [comment]
It seems I have reproduced it in the multi-threaded config

**Comment by civodul** [comment]
Ivan Maidanski <notifications@github.com> skribis:

> It seems I have reproduced it in the multi-threaded config

Yes, that was in a multi-threaded config on x86_64-linux-gnu.

Ludo’.

**Comment by ivmai** [comment]
Should be fixed by b04b346. Please test.
